soccer/src/soccer/strategy/agent/position/offense.hpp

```diff
@@ -29,102 +29,205 @@ class Offense : public Position {
 
     enum State {
-        IDLING,              // simply staying in place
-        SEARCHING,           // moving around on the field to get open
-        PASSING,             // physically kicking the ball towards another robot
-        PREPARING_SHOT,      // pivot around ball in preparation for shot
-        SHOOTING,            // physically kicking the ball towards the net
-        RECEIVING,           // physically intercepting the ball from a pass (gets possession)
-        STEALING,            // attempting to intercept the ball from the other team
-        FACING,              // turning to face the ball
-        SCORER,              // overrides everything and will attempt to steal the bal and shoot it
-        AWAITING_SEND_PASS,  // is waiting to send a pass to someone else
-        SEEKING,             // is trying to get open
+        DEFAULT,           // Decide what to do
+        SEEKING_START,     // Calculate seeking point
+        SEEKING,           // Get open
+        POSSESSION_START,  // Try to shoot and send pass request
+        POSSESSION,        // Holding the ball
+        PASSING_START,     // Prepare to pass
+        PASSING,           // Getting rid of it
+        STEALING,          // Getting the ball
+        RECEIVING_START,   // Facing the ball
+        RECEIVING,         // Getting the ball from a pass
+        SHOOTING_START,    // Calculate shot
+        SHOOTING,          // Winning the game
     };
 
-    State update_state();
+    /**
+     * @return what the state should be right now. called on each get_task tick
+     */
+    State next_state();
 
+    /**
+     * @return the task to execute. called on each get_task tick AFTER next_state()
+     */
     std::optional<RobotIntent> state_to_task(RobotIntent intent);
 
-    // current state of the offensive agent (state machine)
-    State current_state_ = IDLING;
-
-    bool scorer_ = false;
-    bool last_scorer_ = false;
-
-    bool has_open_shot();
-
     /**
-     * @brief Send request to the other robots to see if this robot should be the scorer
+     * @brief This FSM has timeouts for certain states.
+     * Ideally, these would not be necessary; as planners get more sophisticated
+     * they should not get "stuck". However, empirically, the offense FSM in particular
+     * has been observed to deadlock often.
+     *
+     * One common case is when waiting for a receiver to accept a pass; if no receiver responds,
+     * the timeout is necessary. In the future receivers may be able to respond in the negative
+     * instead of ignoring the request.
+     *
+     * The timeouts are a safety mechanism, and should not be the primary reason for a
+     * state transition. They are set relatively high for this reason.
+     *
+     * @return the maximum duration to stay in a given state, or -1 if there is no maximum.
      *
      */
-    void send_scorer_request();
+    static constexpr RJ::Seconds timeout(State s) {
```

Comment:
should this many states be infinite?

Reply:
Looks like we're calling timed_out() in POSSESSION and POSSESSION_START. So they shouldn't be infinite. What values are we thinking (basically, how long after possession should we give up on a pass response and just shoot?)